framework/src/systems/SystemBase.C

```diff
@@ -738,10 +738,34 @@ SystemBase::addVariable(const std::string & var_type,
     if (fe_field_type == TYPE_VECTOR)
       mooseError("Vector family type cannot be used in an array variable");
 
+    std::vector<std::string> array_var_component_names;
+    const bool has_array_names = parameters.isParamValid("array_var_component_names");
+    if (has_array_names)
+    {
+      array_var_component_names =
+          parameters.get<std::vector<std::string>>("array_var_component_names");
+      if (array_var_component_names.size() != components)
```

Comment:
Can't you keep this error within `MooseVariableBase`? That way we can use `paramError()` there and get context by line to the offending parameter?

Reply:
The difficulty with how this works is that SystemBase needs the parameters before the MooseVariableBase object is constructed. That is why the check of the `array` parameter was also here in SystemBase instead of MooseVariableBase. But I could of course find some way to indicate to MooseVariableBase that `array_var_component_names` was invalid and it could throw a paramError instead of SystemBase throwing a mooseError. I think the current code is a little cleaner even though it is using a mooseError where a paramError would be more appropriate. If you look at the error message, it does at least specify which variable has the incorrect parameter so it should be easy for a user to fix the issue.

Reply:
Perhaps we could pass `MooseVariableBase` a system pointer and let it actually add the variable so that it can also check its own parameters? That would be a bigger change, but might make more sense than the way things are now where all the parameters are needed before the `MooseVariableBase` object exists.

Reply:
@loganharbour what is your opinion about this, should I do some more refactoring beyond the current changes so `paramError` can be called from `MooseVariableBase`?

Reply:
Just leave it as is. We can add more context to this error in the future